webrender/src/frame.rs

```diff
@@ -296,6 +298,40 @@ impl Frame {
     }
 
     /// Returns true if any layers actually changed position or false otherwise.
+    pub fn scroll_layers(&mut self,
+                         origin: Point2D<f32>,
+                         pipeline_id: PipelineId,
+                         scroll_root_id: ServoScrollRootId)
+                          -> bool {
+        let origin = LayerPoint::new(origin.x.max(0.0), origin.y.max(0.0));
+
+        let mut scrolled_a_layer = false;
+        let mut found_layer = false;
+        for (layer_id, layer) in self.layers.iter_mut() {
+            if layer_id.pipeline_id != pipeline_id {
+                continue;
+            }
+
+            match layer_id.info {
+                ScrollLayerInfo::Scrollable(_, id) if id != scroll_root_id => continue,
+                ScrollLayerInfo::Fixed => continue,
+                _ => {}
+            }
+
+            found_layer = true;
+            scrolled_a_layer = layer.set_scroll_origin(&origin);
+        }
+
+        if !found_layer {
+            let scroll_offsets =
+                self.pending_scroll_offsets.entry(pipeline_id).or_insert(HashMap::new());
```

Comment:
nit: `or_insert_with(HashMap::new)` doesn't call `HashMap::new()` unconditionally, though I don't know if that allocates so it could be meaningless.